pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -43,11 +43,24 @@
   // Map from MV column name to max number of entries for the column.
   private final Map<String, Integer> _columnToMaxNumMultiValuesMap;
 
-  // Re-usable array for all dictionary ids in the block, of a single valued column
-  private final int[] _reusableDictIds;
+  // Thread local (reusable) array for all dictionary ids in the block, of a single valued column
+  private static final ThreadLocal<int[]> THREAD_LOCAL_DICT_IDS = new ThreadLocal<int[]>() {
+    @Override
+    protected int[] initialValue() {
+      return new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+  };
 
   // Re-usable array to store MV dictionary id's for a given docId
-  private final int[] _reusableMVDictIds;
+  private static final ThreadLocal<int[]> THREAD_LOCAL_MV_DICT_IDS = new ThreadLocal<int[]>() {
```

Comment:
This is only for MV columns. Even if we had an MV column with 1k values (that is already too large), we will use 4k bytes for memory. So I am not convinced this can cause an issue.